From c7e4cb9a8ef2526f0e8414a7c3d441fa918786ae Mon Sep 17 00:00:00 2001
From: Dhina17 <dhinalogu@gmail.com>
Date: Sun, 9 Feb 2025 17:21:53 +0100
Subject: [PATCH 4/9] sweet: Implement torch light control

[RealJohnGalt:] modify for oneplus usage, add toggle switch
[cyberknight777]: Adapt nodes and SEPolicy for sunny

Change-Id: I0b86fbeefbe3b951c5b9827a06c066dd71ada002
Signed-off-by: Cyber Knight <cyberknight755@gmail.com>
Signed-off-by: clarencelol <clarencekuiek@proton.me>
---
 camera/Android.bp                  | 22 ++++++++++++
 camera/CameraProviderExtension.cpp | 65 +++++++++++++++++++++++++++++++++++
 device.mk                          |  3 ++
 3 files changed, 90 insertions(+)
 create mode 100644 camera/Android.bp
 create mode 100644 camera/CameraProviderExtension.cpp

diff --git a/camera/Android.bp b/camera/Android.bp
new file mode 100644
index 0000000..960f03f
--- /dev/null
+++ b/camera/Android.bp
@@ -0,0 +1,22 @@
+//
+// Copyright (C) 2022 The LineageOS Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+cc_library_static {
+    name: "libcameraservice_extension.sweet",
+    srcs: ["CameraProviderExtension.cpp"],
+    include_dirs: [
+        "frameworks/av/services/camera/libcameraservice/common"
+    ],
+}
diff --git a/camera/CameraProviderExtension.cpp b/camera/CameraProviderExtension.cpp
new file mode 100644
index 0000000..95dd68e
--- /dev/null
+++ b/camera/CameraProviderExtension.cpp
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2024 LibreMobileOS Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "CameraProviderExtension.h"
+
+#include <fstream>
+
+#define TORCH_BRIGHTNESS "brightness"
+#define TORCH_MAX_BRIGHTNESS "max_brightness"
+#define TOGGLE_SWITCH "/sys/devices/platform/soc/c440000.qcom,spmi/spmi-0/spmi0-05/c440000.qcom,spmi:qcom,pm6150l@5:qcom,leds@d300/leds/led:switch_2/brightness"
+
+static std::string kTorchLedPaths[] = {
+        "/sys/devices/platform/soc/c440000.qcom,spmi/spmi-0/spmi0-05/c440000.qcom,spmi:qcom,pm6150l@5:qcom,leds@d300/leds/led:torch_0",
+        "/sys/devices/platform/soc/c440000.qcom,spmi/spmi-0/spmi0-05/c440000.qcom,spmi:qcom,pm6150l@5:qcom,leds@d300/leds/led:torch_1",
+};
+
+/** Write value to path and close file. */
+template <typename T>
+static void set(const std::string& path, const T& value) {
+    std::ofstream file(path);
+    file << value;
+}
+
+/** Read value from the path and close file. */
+template <typename T>
+static T get(const std::string& path, const T& def) {
+    std::ifstream file(path);
+    T result;
+
+    file >> result;
+    return file.fail() ? def : result;
+}
+
+bool supportsTorchStrengthControlExt() {
+    return true;
+}
+
+int32_t getTorchDefaultStrengthLevelExt() {
+    return 7;
+}
+
+int32_t getTorchMaxStrengthLevelExt() {
+    // In our device, both LEDs has same maximum value so get from one.
+    auto node = kTorchLedPaths[0] + "/" + TORCH_MAX_BRIGHTNESS;
+    return get(node, 0);
+}
+
+int32_t getTorchStrengthLevelExt() {
+    // We write same value in both LEDs, so get from one.
+    auto node = kTorchLedPaths[0] + "/" + TORCH_BRIGHTNESS;
+    return get(node, 0);
+}
+
+void setTorchStrengthLevelExt(int32_t torchStrength) {
+    set(TOGGLE_SWITCH, 0);
+    for (auto& path : kTorchLedPaths) {
+        auto node = path + "/" + TORCH_BRIGHTNESS;
+        set(node, torchStrength);
+    }
+    if (torchStrength > 0)
+        set(TOGGLE_SWITCH, 255);
+}
diff --git a/device.mk b/device.mk
index 1bb35ff..0d2d4f2 100644
--- a/device.mk
+++ b/device.mk
@@ -21,6 +21,9 @@ PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/configs/audio/mixer_paths_overlay_static.xml:$(TARGET_COPY_OUT_VENDOR)/etc/mixer_paths_overlay_static.xml \
     $(LOCAL_PATH)/configs/audio/sound_trigger_mixer_paths.xml:$(TARGET_COPY_OUT_VENDOR)/etc/sound_trigger_mixer_paths.xml
 
+# Camera
+$(call soong_config_set,libcameraservice,ext_lib,//$(LOCAL_PATH):libcameraservice_extension.sweet)
+
 # Display
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/configs/display/qdcm_calib_data_xiaomi_k6_38_0c_0a_fhd_dsc_video_dsi_panel.xml:$(TARGET_COPY_OUT_VENDOR)/etc/qdcm_calib_data_xiaomi_k6_38_0c_0a_fhd_dsc_video_dsi_panel.xml \
-- 
2.53.0
